client-sdk/go/types/transaction.go

```diff
@@ -419,9 +489,96 @@ type PrettyTransaction struct {
 	AuthInfo AuthInfo   `json:"ai"`
 }
 
+func (ptx *PrettyTransaction) PrettyPrint(ctx context.Context, prefix string, w io.Writer) {
+	fmt.Fprintf(w, "%sFormat: %s\n", prefix, ptx.Call.Format)
+
+	if ptx.Call.Method != "" {
+		fmt.Fprintf(w, "%sMethod: %s\n", prefix, ptx.Call.Method)
+	}
+
+	fmt.Fprintf(w, "%sBody:\n", prefix)
+	// If the body type supports pretty printing, use that.
+	if pp, ok := ptx.Call.Body.(prettyprint.PrettyPrinter); ok {
+		pp.PrettyPrint(ctx, prefix+"  ", w)
+	} else {
+		// Otherwise, just serialize into JSON and display that.
+		data, err := json.MarshalIndent(ptx.Call.Body, prefix+"  ", "  ")
+		if err != nil {
+			fmt.Fprintf(w, "%s<error: %s>\n", prefix+"  ", err)
+		}
+		fmt.Fprintf(w, "%s%s\n", prefix+"  ", data)
+	}
+
+	fmt.Fprintf(w, "%sAuthorized signer(s):\n", prefix)
+	for idx, si := range ptx.AuthInfo.SignerInfo {
+		fmt.Fprintf(w, "%s%d. %s (%s)\n", prefix, idx+1, si.AddressSpec.Signature.PublicKey(), prettyAddressSpecSig(si.AddressSpec.Signature))
+		fmt.Fprintf(w, "%s   Nonce: %d\n", prefix, si.Nonce)
+	}
+
+	fee, err := json.MarshalIndent(ptx.AuthInfo.Fee, prefix+"  ", "  ")
+	if err != nil {
+		fmt.Fprintf(w, "%s<error: %s>\n", prefix+"  ", err)
+	}
+	fmt.Fprintf(w, "%sFee:\n", prefix)
+	fmt.Fprintf(w, "%s%s\n", prefix+"  ", fee)
+}
+
+func prettyAddressSpecSig(spec interface{}) string {
+	specMap := map[string]interface{}{}
+
+	specBytes, _ := json.Marshal(spec)
+	_ = json.Unmarshal(specBytes, &specMap)
+
+	for k := range specMap {
+		return k
+	}
+	return ""
+}
+
+func (ptx *PrettyTransaction) PrettyType() (interface{}, error) {
+	return ptx, nil
+}
+
 // PrettyCall returns a representation of the type that can be used for pretty printing.
 type PrettyCall struct {
 	Format CallFormat  `json:"format,omitempty"`
-	Method string      `json:"method,omitempty"`
+	Method MethodName  `json:"method,omitempty"`
 	Body   interface{} `json:"body"`
 }
+
+// MethodName is a method name.
+type MethodName string
+
+// SanityCheck performs a basic sanity check on the method name.
+func (m MethodName) SanityCheck() error {
+	if len(m) == 0 {
+		return fmt.Errorf("transaction: empty method")
+	}
+
+	return nil
+}
+
+// BodyType returns the registered body type associated with this method.
+func (m MethodName) BodyType() interface{} {
+	bodyType, _ := registeredMethods.Load(string(m))
+	return bodyType
+}
+
+// NewMethodName creates a new method name.
+//
+// Module and method pair must be unique. If they are not, this method
+// will panic.
+func NewMethodName(name string, bodyType interface{}) MethodName {
+	// Check for duplicate method names.
+	if _, isRegistered := registeredMethods.Load(name); isRegistered {
+		panic(fmt.Errorf("transaction: method already registered: %s", name))
+	}
+	registeredMethods.Store(name, bodyType)
+
+	return MethodName(name)
+}
+
+// MethodNames returns a map of all registered method names and its class instances.
+func MethodNames() *sync.Map {
```

Comment:
Why is this needed? In any case it should probably create a copy (e.g. in a regular map) instead of exposing the sync map directly as this allows it to be mutated etc.